src/commands/moderation/mod.rs

```diff
@@ -1,3 +1,140 @@
+use crate::Context;
+use std::str::FromStr;
+
+use color_eyre::eyre::Result;
+use poise::serenity_prelude::{User, UserId};
+
 mod actions;
+use actions::{Ban, Kick, ModAction};
+
+fn split_argument<T>(list: String) -> Vec<T>
+where
+    T: FromStr,
+{
+    list.split(',')
+        .filter_map(|s| s.trim().parse().ok())
+        .collect()
+}
+
+/// Ban a user
+#[poise::command(
+    slash_command,
+    prefix_command,
+    default_member_permissions = "BAN_MEMBERS",
+    required_permissions = "BAN_MEMBERS",
+    aliases("ban")
+)]
+pub async fn ban_user(
+    ctx: Context<'_>,
+    #[description = "User to ban"] user: User,
+    #[description = "Reason to ban"] reason: Option<String>,
+    #[description = "Number of days to purge their messages from (defaults to 0)"]
+    purge_messages_days: Option<u8>,
+    #[description = "If true, the reply from the bot will be ephemeral"] quiet: Option<bool>,
+) -> Result<()> {
+    let dmd = purge_messages_days.unwrap_or(0);
+
+    let action = ModAction {
+        reason,
+        data: Ban {
+            purge_messages_days: dmd,
+        },
+    };
+
+    action.handle(&ctx, &user, quiet, true).await?;
+
+    Ok(())
+}
+
+// Ban multiple users
+#[poise::command(
+    slash_command,
+    prefix_command,
+    default_member_permissions = "BAN_MEMBERS",
+    required_permissions = "BAN_MEMBERS",
+    aliases("ban_multi")
+)]
+pub async fn mass_ban(
+    ctx: Context<'_>,
+    #[description = "Comma separated list of users to ban"] users: String,
+    #[description = "Reason to ban"] reason: Option<String>,
+    #[description = "Number of days to purge their messages from (defaults to 0)"]
+    purge_messages_days: Option<u8>,
+    #[description = "If true, the reply from the bot will be ephemeral"] quiet: Option<bool>,
+) -> Result<()> {
+    let dmd = purge_messages_days.unwrap_or(0);
+    let users: Vec<UserId> = split_argument(users);
+
+    for user in &users {
+        let user = ctx.http().get_user(user.0).await?;
+        let action = ModAction {
+            reason: reason.clone(),
+            data: Ban {
+                purge_messages_days: dmd,
+            },
+        };
+
+        action.handle(&ctx, &user, quiet, false).await?;
+    }
```

Comment:
it's pretty annoying to use this as a slash command (and adds more complexity) so I'd suggest disabling that and using a vector! (same for mass kick)

```suggestion
// Ban multiple users
#[poise::command(
    prefix_command,
    default_member_permissions = "BAN_MEMBERS",
    required_permissions = "BAN_MEMBERS",
    aliases("ban_multi")
)]
pub async fn mass_ban(
    ctx: Context<'_>,
    #[description = "Users to ban"] users: Vec<User>,
    #[description = "Reason to ban"] reason: Option<String>,
    #[description = "Number of days to purge their messages from (defaults to 0)"]
    purge_messages_days: Option<u8>,
    #[description = "If true, the reply from the bot will be ephemeral"] quiet: Option<bool>,
) -> Result<()> {
    let dmd = purge_messages_days.unwrap_or(0);

    for user in &users {
        let user = ctx.http().get_user(user.0).await?;
        let action = ModAction {
            reason: reason.clone(),
            data: Ban {
                purge_messages_days: dmd,
            },
        };

        action.handle(&ctx, &user, quiet, false).await?;
    }
```

Reply:
this was intentional; Vec<User> doesn't work with slash commands and will only accept a string with no autocomplete afterwards, so it's either a main ban command that has the the autocomplete expected or none. the multi-* commands do suck with slash though because of this limitation, so i'd probably just recommend people use the prefix command for the few times we'll run it

i also just don't see much added complexity from having a separate command when like i said before, it'll be rare for us to use it and it's pretty obvious in it's function